File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Tests/Api/CustomerServiceNegativeTest.cs
using System;
using System.Collections.Generic;
using System.Net;
using FluentAssertions;
using NUnit.Extension.DependencyInjection;
using RekindleTaf.Controllers;
using RekindleTaf.Exceptions;
using RekindleTaf.Model;
using RekindleTaf.Model.Factory;
using RekindleTaf.Utils;

namespace RekindleTaf.Tests.Api;
[DependencyInjectingTestFixture]
public class CustomerServiceNegativeTest(IFactoryKit factoryKit, ICustomerController customerController) : Base
{
    [Test]
    public void Verify_CannotCreateCustomer_WithEmptyValues()
    {
        var customer = new Customer{FirstName = "", LastName = "", Username = ""};
        var response = customerController.Post(customer).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("firstName").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("lastName").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("username").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
    }
    
    [Test]
    public void Verify_CannotCreateCustomer_WithNullValues()
    {
        var customer = new Customer();
        var response = customerController.Post(customer).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("firstName").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("lastName").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("username").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
    }
    
    [Test]
    public void Verify_CannotGet_NonExistingCustomer()
    {
        var response = customerController.Get(Guid.NewGuid().ToString()).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.CustomerNotFound);
    }
    
    [Test]
    public void Verify_CannotDelete_NonExistingCustomer()
    {
        var customer = new Customer {Id = Guid.NewGuid().ToString()};
        var response = customerController.Delete(customer).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.CustomerNotFound);
    }
    
    [Test]
    public void Verify_CannotUpdate_NonExistingCustomer()
    {
        var customer = factoryKit.Factory<Customer>(FactoryKey.Customer).Create(CustomerKey.StdUser);
        customer.Id = Guid.NewGuid().ToString();
        var response = customerController.Put(customer).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.CustomerNotFound);
    }
}

File: typescript-project/rekindle-book-store-aqa/tests/customer-service-negative.test.ts
import {beforeAll, describe, expect, test} from "@jest/globals";
import {Customer} from "@/model/domain-model";
import {Factory, FactoryKey, FactoryKit} from "@/model/factory";
import {CustomerController} from "@/rest/controllers/customer-controller";
import {CustomerKey} from "@/model/customer-factory";
import {AxiosResponse, HttpStatusCode} from "axios";
import {ErrorMessage} from "../src/utils/utils";
import {container} from "tsyringe";
import {RestClient} from "@/rest/rekindle-rest-client";
//@ts-ignore
import {v4 as uuidv4} from 'uuid';

describe('Customer Service Test', () => {
    let restClient: RestClient = container.resolve('RestClient');
    let factoryKit: FactoryKit = container.resolve('FactoryKit');
    let customerController: CustomerController = container.resolve('CustomerCrudController');
    let customerFactory: Factory<Customer> = factoryKit.factory<Customer>(FactoryKey.customer);

    beforeAll(async () => {
        await restClient.setupToken();
    });

    test('should not create customer with empty values', async () => {
        const customer: Customer = {firstName: '', lastName: '', username: ''};
        await customerController.post(customer).catch((error) => {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.firstName);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.lastName);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.username);
        })
    });

    test('should not create customer with null values', async () => {
        const customer: Customer = {};
        await customerController.post(customer).catch((error) => {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.firstName);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.lastName);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.username);
        })
    });

    test('should not get non-existing customer', async () => {
        await customerController.get(uuidv4().toString()).catch((error) => {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.customerNotFound);
        });
    });

    test('should not delete non-existing customer', async () => {
        const customer: Customer = {id: uuidv4().toString()};
        await customerController.delete(customer).catch((error) => {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.customerNotFound);
        });
    });

    test('should not update non-existing customer', async () => {
        const customer = customerFactory.create(CustomerKey.stdCustomerActive);
        customer.id = uuidv4().toString();

        await customerController.put(customer).catch((error) => {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.customerNotFound);
        })
    });
});